### PR TITLE
SONARJAVA-1319 Make parent link available through tree interface

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/resolve/SemanticModel.java [hunk 1]
```diff
@@ -137,19 +120,14 @@ public Tree getTree(Resolve.Env env) {
...
 
   public Resolve.Env getEnv(Tree tree) {
-    JavaTree javaTree = (JavaTree) tree;
     Resolve.Env result = null;
-    while (result == null && javaTree != null) {
-      result = envs.get(javaTree);
-      javaTree = (JavaTree) parentLink.get(javaTree);
+    while (result == null && tree != null) {
+      result = envs.get(tree);
+      tree = tree.parent();
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/SemanticModel.java]: MAJOR Introduce a new variable instead of reusing the parameter "tree". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1226)

File: java-squid/src/main/java/org/sonar/java/ast/parser/JavaParser.java [hunk 2]
```diff
@@ -39,4 +45,26 @@ private JavaParser() {
...
+  public Tree parse(File file) {
+    return createParentLink((JavaTree) super.parse(file));
+  }
+
+  @Override
+  public Tree parse(String source) {
+    return createParentLink((JavaTree) super.parse(source));
+  }
+
+  private Tree createParentLink(JavaTree parent) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/JavaParser.java]: MAJOR Make "createParentLink" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-squid/src/test/java/org/sonar/java/ast/parser/JavaParserTest.java [hunk 3]
```diff
@@ -17,38 +17,24 @@
...
-    VariableTree field = (VariableTree) classTree.members().get(0);
-    MethodTree method = (MethodTree) classTree.members().get(1);
-
-    assertThat(semanticModel.getParent(method)).isSameAs(classTree);
-    assertThat(semanticModel.getParent(field)).isSameAs(classTree);
-    assertThat(semanticModel.getParent(classTree)).isSameAs(cut);
-
+    MethodTree method = (MethodTree) classTree.members().get(0);
+    assertThat(method.parent()).isSameAs(classTree);
+    assertThat(classTree.parent()).isSameAs(cut);
```
Comment (m-g-sonar) [inline java-squid/src/test/java/org/sonar/java/ast/parser/JavaParserTest.java]: I would add a test asserting that the parent of the `CompilationUnitTree` is `null`.

Comment (benzonico) [inline java-squid/src/test/java/org/sonar/java/ast/parser/JavaParserTest.java]: agreed ! fixed.

File: java-squid/src/main/java/org/sonar/java/ast/parser/JavaParser.java [hunk 4]
```diff
@@ -39,4 +45,26 @@ private JavaParser() {
...
+  public Tree parse(String source) {
+    return createParentLink((JavaTree) super.parse(source));
+  }
+
+  private static Tree createParentLink(JavaTree parent) {
+    if (!parent.isLeaf()) {
+      for (Iterator<Tree> iter = parent.childrenIterator(); iter.hasNext();) {
+        JavaTree next = (JavaTree) iter.next();
+        if (next != null) {
+          next.setParent(parent);
```
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/ast/parser/JavaParser.java]: So, for trees "sharing" children (which is already known to be problematic on our side), we will end up with a single parent all the time, which is not a real representation of the hierarchy...

With the following code ```final int a = 0, b = 0;```, I would have expected the `ModifierKeywordTree` corresponding to `final`, or the `TypeTree` corresponding to `int`, to have two parents.

Consequently, if you try something like in the following code snippet, you will face strange behavior. Given the `VariableTree` b.

```
VariableTree a = b.modifiers().parent(); // when you would expect to get back to b
```

Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/ast/parser/JavaParser.java]: This is a known issue but not relevant to this ticket : https://jira.sonarsource.com/browse/SONARJAVA-1176
